File: packages/sdk-java/src/main/java/ee/apexmail/APIKeys.java
package ee.apexmail;

import com.fasterxml.jackson.core.type.TypeReference;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.Map;

/** Manage API keys for the authenticated account. */
public final class APIKeys {
    private final ApexMailClient client;

    public APIKeys(ApexMailClient client) {
        this.client = client;
    }

    /**
     * Create a new API key.
     *
     * <p>Transmits exactly {name, scopes: string[], expires_in_days?} per
     * the server's CreateApiKeyRequest (auth.rs, deny_unknown_fields) —
     * {@code scopes} is required by the API and defaults to an empty list.
     * The legacy {@code expiresAt} input is ignored; use
     * {@code expires_in_days} (1..365).
     */
    public Map<String, Object> create(Map<String, Object> params) {
        if (params == null || params.get("name") == null || String.valueOf(params.get("name")).isBlank()) {
            throw new IllegalArgumentException("name is required");
        }
        Map<String, Object> body = new java.util.LinkedHashMap<>();
        body.put("name", params.get("name"));
        Object scopes = params.get("scopes");
        body.put("scopes", scopes != null ? scopes : java.util.List.of());
        if (params.get("expires_in_days") != null) {
            body.put("expires_in_days", params.get("expires_in_days"));
        }
        return client.request("POST", "/v1/auth/api-keys", body, new TypeReference<Map<String, Object>>() {});
    }

    public Map<String, Object> list(Map<String, Object> options) {
        String query = options != null && !options.isEmpty() ? "?" + buildQuery(options) : "";
        return client.request("GET", "/v1/auth/api-keys" + query, null, new TypeReference<Map<String, Object>>() {});
    }

    public Map<String, Object> list() {
        return list(Map.of("limit", 50, "offset", 0));
    }

    public void revoke(String id) {
        client.request("DELETE", "/v1/auth/api-keys/" + encode(id), null, Void.class);
    }

    private static String buildQuery(Map<String, Object> params) {
        StringBuilder builder = new StringBuilder();
        params.forEach((key, value) -> {
            if (value != null) {
                if (builder.length() > 0) builder.append('&');
                builder.append(encode(key)).append('=').append(encode(String.valueOf(value)));
            }
        });
        return builder.toString();
    }

    private static String encode(Object value) {
        return URLEncoder.encode(String.valueOf(value), StandardCharsets.UTF_8);
    }
}

File: packages/sdk-php/src/Resources/ApiKeys.php
<?php

declare(strict_types=1);

namespace ApexMail\Resources;

use ApexMail\Client;

/**
 * Manage API keys for the authenticated account.
 */
class ApiKeys
{
    public function __construct(private readonly Client $client) {}

    /**
     * Create a new API key.
     *
     * The server's CreateApiKeyRequest accepts exactly
     * {name, scopes: string[], expires_in_days?} (deny_unknown_fields).
     * `scopes` is required by the API and defaults to [] (a key with no
     * scopes); the legacy expiresAt input is accepted but not sent — use
     * expires_in_days (1..365).
     *
     * @param array $params { name, scopes?, expires_in_days?, expires_at? (unused) }
     */
    public function create(array $params): array
    {
        $scopes = $params['scopes'] ?? [];
        $body = array_filter([
            'name'             => $params['name'] ?? null,
            'scopes'           => array_values(array_map('strval', (array) $scopes)),
            'expires_in_days'  => $params['expires_in_days'] ?? $params['expiresInDays'] ?? null,
        ], static fn ($value) => $value !== null && $value !== '');

        // scopes must always be present (even empty) for the server DTO.
        if (!isset($body['scopes'])) {
            $body['scopes'] = [];
        }

        return $this->client->request('POST', '/v1/auth/api-keys', $body);
    }

    /**
     * List API keys.
     *
     * @param array $options { limit, offset }
     */
    public function list(array $options = []): array
    {
        $query = http_build_query(array_filter([
            'limit'  => $options['limit']  ?? 50,
            'offset' => $options['offset'] ?? 0,
            'cursor' => $options['cursor'] ?? null,
        ], static fn ($value) => $value !== null && $value !== ''));

        return $this->client->request('GET', '/v1/auth/api-keys' . ($query ? '?' . $query : ''));
    }

    /**
     * Revoke an API key.
     */
    public function revoke(string $id): array
    {
        return $this->client->request('DELETE', '/v1/auth/api-keys/' . urlencode($id));
    }
}